clang/lib/Sema/BoundsWideningAnalysis.cpp

```diff
@@ -14,4 +14,1213 @@
 #include "clang/Sema/BoundsWideningAnalysis.h"
 
 namespace clang {
+
+//===---------------------------------------------------------------------===//
+// Implementation of the methods in the BoundsWideningAnalysis class. This is
+// the main class that implements the dataflow analysis for bounds widening of
+// null-terminated arrays. This class uses helper methods from the
+// BoundsWideningUtil class that are defined later in this file.
+//===---------------------------------------------------------------------===//
+
+void BoundsWideningAnalysis::WidenBounds(FunctionDecl *FD,
+                                         StmtSetTy NestedStmts) {
+  assert(Cfg && "expected CFG to exist");
+
+  // Initialize the list of variables that are pointers to null-terminated
+  // arrays. This list will be initialized with the variables that are passed
+  // as parameters to the function.
+  InitNtPtrsInFunc(FD);
+
+  // WorkList will store the blocks that remain to be processed for the
+  // fixedpoint computation.
+  WorkListTy WorkList;
+
+  // Note: By default, PostOrderCFGView iterates in reverse order. So we always
+  // get a reverse post order when we iterate PostOrderCFGView.
+  for (const CFGBlock *B : PostOrderCFGView(Cfg)) {
+    // SkipBlock will skip all null blocks and the exit block. PostOrderCFGView
+    // does not traverse any unreachable blocks. So at the end of this loop
+    // BlockMap only contains reachable blocks.
+    if (BWUtil.SkipBlock(B))
+      continue;
+
+    // Create a mapping from CFGBlock to ElevatedCFGBlock.
+    auto EB = new ElevatedCFGBlock(B);
+    BlockMap[B] = EB;
+
+    // Note: WorkList is a queue. So we maintain the reverse post order when we
+    // iterate WorkList.
+    WorkList.append(EB);
+
+    // Compute Gen and Kill sets for the block and statements in the block.
+    ComputeGenKillSets(EB, NestedStmts);
+
+    // Initialize the In and Out sets for the block.
+    InitBlockInOutSets(FD, EB);
+  }
+
+  // Compute the In and Out sets for blocks. This is the fixedpoint computation
+  // for the dataflow analysis.
+  while (!WorkList.empty()) {
+    ElevatedCFGBlock *EB = WorkList.next();
+    WorkList.remove(EB);
+
+    ComputeInSet(EB);
+    ComputeOutSet(EB, WorkList);
```

Comment:
If possible, the code for adding the successors of `EB` into the `WorkList` can be moved here to make the fixpoint computation clearer. For example, `ComputeOutSet` can return a list (or set) of successor basic blocks that need to be added to `WorkList` (this list can be empty if the out set is unchanged).
If this is hard, then there should be a clear comment here explaining this.